File: ttshare_mobile/lib/services/webdav_client.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class WebdavClient {
  String? _baseUrl;
  String? _username;
  String? _password;

  bool get isConfigured =>
      _baseUrl != null && _username != null && _password != null;

  String? get baseUrl => _baseUrl;

  Future<void> loadConfig() async {
    final prefs = await SharedPreferences.getInstance();
    _baseUrl = prefs.getString('webdav_url');
    _username = prefs.getString('webdav_username');
    _password = prefs.getString('webdav_password');
  }

  Future<void> saveConfig(String url, String username, String password) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('webdav_url', url);
    await prefs.setString('webdav_username', username);
    await prefs.setString('webdav_password', password);
    _baseUrl = url;
    _username = username;
    _password = password;
  }

  Future<bool> verifyConnection() async {
    try {
      final uri = Uri.parse(_baseUrl!);
      final response = await http.get(uri, headers: _authHeaders);
      return response.statusCode == 200 || response.statusCode == 207;
    } catch (e) {
      return false;
    }
  }

  Map<String, String> get _authHeaders {
    final bytes = utf8.encode('$_username:$_password');
    return {'Authorization': 'Basic ${base64.encode(bytes)}'};
  }

  Future<void> createFolder(String path) async {
    final uri = Uri.parse('$_baseUrl$path');
    final request = http.Request('MKCOL', uri);
    request.headers.addAll(_authHeaders);
    final streamed = await request.send();
    final status = streamed.statusCode;
    // 201=created, 200=ok, 301=redirect, 405=already exists
    if (status != 201 && status != 200 && status != 301 && status != 405) {
      throw Exception('Failed to create folder: $status');
    }
  }

  Future<void> uploadFile(String localPath, String remotePath) async {
    final uri = Uri.parse('$_baseUrl$remotePath');
    final file = File(localPath);
    final bytes = await file.readAsBytes();
    final request = http.Request('PUT', uri);
    request.headers.addAll(_authHeaders);
    request.headers['Content-Type'] = 'text/html; charset=utf-8';
    request.bodyBytes = bytes;
    final streamed = await request.send();
    if (streamed.statusCode != 201 && streamed.statusCode != 200) {
      throw Exception('Failed to upload file: ${streamed.statusCode}');
    }
  }

  Future<void> deleteFile(String path) async {
    final uri = Uri.parse('$_baseUrl$path');
    final request = http.Request('DELETE', uri);
    request.headers.addAll(_authHeaders);
    final streamed = await request.send();
    if (streamed.statusCode != 204 && streamed.statusCode != 200) {
      throw Exception('Failed to delete file: ${streamed.statusCode}');
    }
  }
}

File: ttshare_android/app/src/main/java/com/ttshare/service/WebdavClient.kt
package com.ttshare.service

import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.util.Base64

class WebdavClient(private val prefs: SharedPreferences) {

    private var baseUrl: String?
        get() = prefs.getString("webdav_url", null)
        set(value) = prefs.edit().putString("webdav_url", value).apply()

    private var username: String?
        get() = prefs.getString("webdav_username", null)
        set(value) = prefs.edit().putString("webdav_username", value).apply()

    private var password: String?
        get() = prefs.getString("webdav_password", null)
        set(value) = prefs.edit().putString("webdav_password", value).apply()

    val isConfigured: Boolean
        get() = baseUrl != null && username != null && password != null

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    private val authHeader: String
        get() {
            val raw = "${username ?: ""}:${password ?: ""}"
            return "Basic ${Base64.getEncoder().encodeToString(raw.toByteArray())}"
        }

    fun saveConfig(url: String, user: String, pass: String) {
        baseUrl = url
        username = user
        password = pass
    }

    suspend fun verifyConnection(): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url(baseUrl ?: return@withContext false)
                .header("Authorization", authHeader)
                .build()
            val response = client.newCall(request).execute()
            response.code in listOf(200, 207, 301, 302)
        } catch (e: Exception) {
            false
        }
    }

    suspend fun createFolder(path: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${baseUrl}$path")
            .method("MKCOL", null)
            .header("Authorization", authHeader)
            .build()
        val response = client.newCall(request).execute()
        if (response.code !in listOf(201, 200, 301, 405)) {
            throw IOException("Failed to create folder: ${response.code}")
        }
    }

    suspend fun uploadFile(localPath: String, remotePath: String) = withContext(Dispatchers.IO) {
        val file = File(localPath)
        val body = file.readBytes().toRequestBody("text/html; charset=utf-8".toMediaType())
        val request = Request.Builder()
            .url("${baseUrl}$remotePath")
            .put(body)
            .header("Authorization", authHeader)
            .build()
        val response = client.newCall(request).execute()
        if (response.code !in listOf(201, 200)) {
            throw IOException("Failed to upload file: ${response.code}")
        }
    }

    suspend fun deleteFile(path: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${baseUrl}$path")
            .delete()
            .header("Authorization", authHeader)
            .build()
        val response = client.newCall(request).execute()
        if (response.code !in listOf(204, 200)) {
            throw IOException("Failed to delete: ${response.code}")
        }
    }
}
